src/Avalonia.Controls.TreeDataGrid/Selection/TreeDataGridRowSelectionModel.cs

```diff
@@ -184,6 +188,85 @@ private void Search(TreeDataGrid treeDataGrid, string candidatePattern, int sele
             }
         }
 
+        void ITreeDataGridSelectionInteraction.OnPreviewKeyDown(TreeDataGrid sender, KeyEventArgs e)
+        {
+            static bool IsElementFullyVisibleToUser(TransformedBounds controlBounds)
+            {
+                return controlBounds.Clip.Contains(controlBounds.Bounds.TransformToAABB(controlBounds.Transform));
+            }
+            if (e.Key == Key.PageDown || e.Key == Key.PageUp)
+            {
+                var children = sender.RowsPresenter.GetLogicalChildren();
+                var childrenCount = children.Count();
+                if (childrenCount > 0)
+                {
+                    if (e.Key == Key.PageDown)
+                    {
+                        if (_lastPageSelectedIndex > 0 && SelectedIndex == _lastPageSelectedIndex && sender.RowsPresenter?.Items != null)
+                        {
+                            if (childrenCount + _lastPageSelectedIndex <= sender.RowsPresenter.Items.Count)
+                            {
+                                _lastPageSelectedIndex = childrenCount - 1 + _lastPageSelectedIndex;
+                            }
+                            else
+                            {
+                                _lastPageSelectedIndex = sender.RowsPresenter.Items.Count - 1;
+                            }
+
+                        }
+                        else if (_lastPageSelectedIndex == 0 || SelectedIndex != _lastPageSelectedIndex)
+                        {
+                            var max = -1;
+                            for (int i = 0; i <= childrenCount - 1; i++)
```

Comment:
If we were to use `RealizedElements` instead of logical children, I don't think we'd need to search all elements here: because `RealizedElements` will be returned in row-index order the list can be searched from the bottom to find the first fully visible element, and exit early once that is found.